WordPress/openverse
# Switch API package management from Pipenv to PDM

## Related

Related to #286. Additional background can be read in [this comment](https://github.com/WordPress/openverse/issues/286#issuecomment-2053604034).

## Description



This PR replaces Pipenv with PDM. While PDM has monorepo support, we have opted not to use this feature in this PR. Instead we will use a path dependency to depend on subpackages located at `packages/` in the monorepo. To see this in action, the license and attribution related functionality in the API has been separated into its own package at `packages/ov-attribution`.

The `ov-attribution` subpackage is listed as a dependency as well as an editable dev-dependency of the API. This means that in dev mode, the package will not actually be copied into the virtualenv but rather only point to `../packages/ov-attribution`. When installed with `--no-editable` it will be copied into the virtualenv. Thanks @sarayourfriend for suggesting this and making usage of PDM possible.

## Testing Instructions


1. Install PDM. I recommend using the `pipx` approach, I am too.
2. Build the new Docker image with `just dc build web`.
3. Explore the editable local package by changing something in `py-packages/openverse-attribution` to see the change immediately in the API.
6. Run the tests. (Or don't, CI has you covered.)
   - Test `openverse-attribution` with:
   ```
   cd py-packages/openverse-attribution && pdm install && pdm run pytest`
   ```
   - Test the API with `just api/test`.

## Checklist



[best_practices]:
  https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin


<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>
